ParallelCascadePlatformTest.testCascadeRaceCondition

``````
cascading.cascade.CascadeException: flow failed: race-50/first-nondeterministic
    at cascading.cascade.BaseCascade$CascadeJob.call(BaseCascade.java:963)
    at cascading.cascade.BaseCascade$CascadeJob.call(BaseCascade.java:900)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: cascading.flow.FlowException: [race-50/first-nondeterministic] unhandled exception
    at cascading.flow.BaseFlow.complete(BaseFlow.java:1012)
    at cascading.cascade.BaseCascade$CascadeJob.call(BaseCascade.java:953)
    at cascading.cascade.BaseCascade$CascadeJob.call(BaseCascade.java:900)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.concurrent.ExecutionException: java.lang.RuntimeException: com.hazelcast.jet.impl.statemachine.InvalidEventException: Invalid Event FINALIZATION_START for state=INTERRUPTION_IN_PROGRESS for stateMachine with name=9993c01100174f9daa7f5abe8e0f93bb
    at java.util.concurrent.FutureTask.report(FutureTask.java:122)
    at java.util.concurrent.FutureTask.get(FutureTask.java:192)
    at cascading.flow.BaseFlow.run(BaseFlow.java:1252)
    at cascading.flow.BaseFlow.access$100(BaseFlow.java:82)
    at cascading.flow.BaseFlow$1.run(BaseFlow.java:928)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: com.hazelcast.jet.impl.statemachine.InvalidEventException: Invalid Event FINALIZATION_START for state=INTERRUPTION_IN_PROGRESS for stateMachine with name=9993c01100174f9daa7f5abe8e0f93bb
    at com.hazelcast.jet.impl.util.JetUtil.reThrow(JetUtil.java:181)
    at com.hazelcast.jet.impl.job.ServerJobInvocation$1.onFailure(ServerJobInvocation.java:52)
    at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:249)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
    at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
    at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)
Caused by: com.hazelcast.jet.impl.statemachine.InvalidEventException: Invalid Event FINALIZATION_START for state=INTERRUPTION_IN_PROGRESS for stateMachine with name=9993c01100174f9daa7f5abe8e0f93bb
    at com.hazelcast.jet.impl.statemachine.StateMachine$EventsProcessor.execute(StateMachine.java:174)
    at com.hazelcast.jet.impl.executor.Worker.execute(Worker.java:140)
    at com.hazelcast.jet.impl.executor.StateMachineWorker.run(StateMachineWorker.java:62)
    at java.lang.Thread.run(Thread.java:745)
    at ------ submitted from ------.(Unknown Source)
    at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:111)
    at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:246)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
    at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
    at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)```
``````

--- a/hazelcast-jet-cascading/src/main/java/com/hazelcast/jet/cascading/planner/JetFlowStepJob.java
+++ b/hazelcast-jet-cascading/src/main/java/com/hazelcast/jet/cascading/planner/JetFlowStepJob.java
@@ -75,7 +75,7 @@ public class JetFlowStepJob extends FlowStepJob<JobConfig> {
                 }
 
                 if (!jobFuture.isDone()) {
-                    job.interrupt();
+                    job.interrupt().get();
                 }
             }
         } catch (Throwable e) {

--- a/hazelcast-jet-cascading/src/test/java/cascading/cascade/ParallelCascadePlatformTest.java
+++ b/hazelcast-jet-cascading/src/test/java/cascading/cascade/ParallelCascadePlatformTest.java
@@ -106,7 +106,6 @@ public class ParallelCascadePlatformTest extends PlatformTestCase
     }
 
   @Test
-  @Ignore //https://github.com/hazelcast/hazelcast-jet/issues/142
   public void testCascadeRaceCondition() throws Throwable
     {
     getPlatform().copyFromLocal( inputFileIps );

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/job/JobClusterService.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/job/JobClusterService.java
@@ -17,7 +17,6 @@
 package com.hazelcast.jet.impl.job;
 
 import com.hazelcast.core.Member;
-import com.hazelcast.jet.CombinedJetException;
 import com.hazelcast.jet.DAG;
 import com.hazelcast.jet.config.DeploymentConfig;
 import com.hazelcast.jet.config.JobConfig;
@@ -27,6 +26,7 @@ import com.hazelcast.jet.impl.job.deployment.ChunkIterator;
 import com.hazelcast.jet.impl.statemachine.job.JobEvent;
 import com.hazelcast.jet.impl.statemachine.job.JobStateMachine;
 import com.hazelcast.jet.impl.util.JetUtil;
+
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
@@ -267,10 +267,7 @@ public abstract class JobClusterService<Payload> {
             }
         }
 
-        if (errors.size() > 1) {
-            throw new CombinedJetException(errors);
-        }
-        if (errors.size() == 1) {
+        if (errors.size() >= 1) {
             throw reThrow(errors.get(0));
         }
     }

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/job/ServerJobInvocation.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/job/ServerJobInvocation.java
@@ -23,9 +23,8 @@ import com.hazelcast.nio.Address;
 import com.hazelcast.spi.InvocationBuilder;
 import com.hazelcast.spi.NodeEngine;
 import com.hazelcast.spi.OperationService;
-import java.util.concurrent.CompletableFuture;
 
-import static com.hazelcast.jet.impl.util.JetUtil.reThrow;
+import java.util.concurrent.CompletableFuture;
 
 public class ServerJobInvocation<T> extends AbstractJobInvocation<JetOperation, T> {
     private final NodeEngine nodeEngine;
@@ -49,7 +48,7 @@ public class ServerJobInvocation<T> extends AbstractJobInvocation<JetOperation, 
 
             @Override
             public void onFailure(Throwable throwable) {
-                completableFuture.completeExceptionally(reThrow(throwable));
+                completableFuture.completeExceptionally(throwable);
             }
         });
         return completableFuture;

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/operation/JobInterruptOperation.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/operation/JobInterruptOperation.java
@@ -18,9 +18,9 @@ package com.hazelcast.jet.impl.operation;
 
 import com.hazelcast.core.ExecutionCallback;
 import com.hazelcast.core.ICompletableFuture;
+import com.hazelcast.jet.impl.job.JobContext;
 import com.hazelcast.jet.impl.runtime.JobManager;
 import com.hazelcast.jet.impl.runtime.jobmanager.JobManagerResponse;
-import com.hazelcast.jet.impl.job.JobContext;
 import com.hazelcast.jet.impl.statemachine.jobmanager.requests.InterruptJobRequest;
 
 public class JobInterruptOperation extends AsyncJetOperation {
@@ -45,6 +45,10 @@ public class JobInterruptOperation extends AsyncJetOperation {
         interruptFuture.andThen(new JobManagerRequestCallback(this,
                 "Unable interrupt job execution", () -> {
             ICompletableFuture<Object> future = jobManager.getInterruptionMailBox();
+            if (future == null) {
+                sendResponse(null);
+                return;
+            }
             future.andThen(new ExecutionCallback<Object>() {
                 @Override
                 public void onResponse(Object response) {

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/runtime/JobManager.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/runtime/JobManager.java
@@ -32,9 +32,7 @@ import com.hazelcast.jet.impl.job.JobContext;
 import com.hazelcast.jet.impl.job.JobException;
 import com.hazelcast.jet.impl.runtime.jobmanager.JobManagerEvent;
 import com.hazelcast.jet.impl.runtime.jobmanager.JobManagerResponse;
-import com.hazelcast.jet.impl.runtime.jobmanager.JobManagerState;
 import com.hazelcast.jet.impl.runtime.task.VertexTask;
-import com.hazelcast.jet.impl.statemachine.StateMachine;
 import com.hazelcast.jet.impl.statemachine.StateMachineRequest;
 import com.hazelcast.jet.impl.statemachine.StateMachineRequestProcessor;
 import com.hazelcast.jet.impl.statemachine.jobmanager.JobManagerStateMachine;
@@ -52,6 +50,7 @@ import com.hazelcast.jet.impl.util.JetUtil;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.nio.Address;
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
@@ -84,7 +83,7 @@ public class JobManager implements StateMachineRequestProcessor<JobManagerEvent>
     private final ILogger logger;
     private final DiscoveryService discoveryService;
     private final JobContext jobContext;
-    private final StateMachine<JobManagerEvent, JobManagerState, JobManagerResponse> stateMachine;
+    private final JobManagerStateMachine stateMachine;
     private volatile boolean interrupted;
     private volatile Throwable interruptionError;
     private volatile DAG dag;
@@ -198,6 +197,7 @@ public class JobManager implements StateMachineRequestProcessor<JobManagerEvent>
         runnerCounter.set(0);
         networkTaskCounter.set(0);
         executionMailBox.set(new BasicCompletableFuture<>(jobContext.getNodeEngine(), logger));
+        interruptionFutureHolder.set(null);
     }
 
     /**
@@ -386,6 +386,10 @@ public class JobManager implements StateMachineRequestProcessor<JobManagerEvent>
      */
     public VertexRunner getRunnerByVertex(Vertex vertex) {
         return vertex2runner.get(vertex);
+    }
+
+    public JobManagerStateMachine getStateMachine() {
+        return stateMachine;
     }
 
     /**

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/statemachine/job/JobStateMachine.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/statemachine/job/JobStateMachine.java
@@ -114,6 +114,8 @@ public class JobStateMachine
                     put(
                             JobState.EXECUTION_SUCCESS, LinkedMapBuilder.of(
                                     JobEvent.EXECUTION_START, JobState.EXECUTION_IN_PROGRESS,
+                                    JobEvent.INTERRUPTION_START, JobState.EXECUTION_SUCCESS,
+                                    JobEvent.INTERRUPTION_SUCCESS, JobState.EXECUTION_SUCCESS,
                                     JobEvent.FINALIZATION_START, JobState.FINALIZATION_IN_PROGRESS,
                                     JobEvent.EXECUTION_SUCCESS, JobState.EXECUTION_SUCCESS,
                                     JobEvent.FINALIZATION_SUCCESS, JobState.NEW

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/statemachine/jobmanager/JobManagerStateMachine.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/statemachine/jobmanager/JobManagerStateMachine.java
@@ -16,14 +16,15 @@
 
 package com.hazelcast.jet.impl.statemachine.jobmanager;
 
+import com.hazelcast.jet.impl.executor.StateMachineExecutor;
+import com.hazelcast.jet.impl.job.JobContext;
 import com.hazelcast.jet.impl.runtime.jobmanager.JobManagerEvent;
 import com.hazelcast.jet.impl.runtime.jobmanager.JobManagerResponse;
 import com.hazelcast.jet.impl.runtime.jobmanager.JobManagerState;
-import com.hazelcast.jet.impl.executor.StateMachineExecutor;
-import com.hazelcast.jet.impl.job.JobContext;
 import com.hazelcast.jet.impl.statemachine.StateMachine;
 import com.hazelcast.jet.impl.statemachine.StateMachineRequestProcessor;
 import com.hazelcast.jet.impl.util.LinkedMapBuilder;
+
 import java.util.Map;
 
 public class JobManagerStateMachine extends StateMachine<JobManagerEvent, JobManagerState, JobManagerResponse> {
@@ -59,7 +60,8 @@ public class JobManagerStateMachine extends StateMachine<JobManagerEvent, JobMan
                     put(
                             JobManagerState.EXECUTION_SUCCESS, LinkedMapBuilder.of(
                                     JobManagerEvent.EXECUTE, JobManagerState.EXECUTING,
-                                    JobManagerEvent.FINALIZE, JobManagerState.FINALIZED
+                                    JobManagerEvent.FINALIZE, JobManagerState.FINALIZED,
+                                    JobManagerEvent.INTERRUPT_EXECUTION, JobManagerState.EXECUTION_SUCCESS
                             )
                     ).
                     put(

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/statemachine/jobmanager/processors/InterruptJobProcessor.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/statemachine/jobmanager/processors/InterruptJobProcessor.java
@@ -17,10 +17,12 @@
 package com.hazelcast.jet.impl.statemachine.jobmanager.processors;
 
 import com.hazelcast.jet.impl.Dummy;
-import com.hazelcast.jet.impl.runtime.VertexRunnerPayloadProcessor;
 import com.hazelcast.jet.impl.runtime.JobManager;
 import com.hazelcast.jet.impl.runtime.VertexRunner;
+import com.hazelcast.jet.impl.runtime.VertexRunnerPayloadProcessor;
+import com.hazelcast.jet.impl.runtime.jobmanager.JobManagerState;
 import com.hazelcast.jet.impl.statemachine.runner.requests.VertexRunnerInterruptRequest;
+
 import java.util.concurrent.TimeUnit;
 
 public class InterruptJobProcessor implements VertexRunnerPayloadProcessor<Dummy> {
@@ -34,6 +36,10 @@ public class InterruptJobProcessor implements VertexRunnerPayloadProcessor<Dummy
 
     @Override
     public void process(Dummy payload) throws Exception {
+        JobManagerState currentState = jobManager.getStateMachine().currentState();
+        if (currentState != JobManagerState.EXECUTING) {
+            return;
+        }
         jobManager.registerInterruption();
 
         for (VertexRunner runner : jobManager.runners()) {

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/statemachine/runner/processors/VertexRunnerInterruptProcessor.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/statemachine/runner/processors/VertexRunnerInterruptProcessor.java
@@ -31,7 +31,6 @@ public class VertexRunnerInterruptProcessor implements VertexRunnerPayloadProces
 
     @Override
     public void process(Throwable error) throws Exception {
-        logger.fine("VertexRunnerInterruptProcessor.process");
         this.vertexRunner.interrupt(error);
     }
 }

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/statemachine/runner/processors/VertexRunnerInterruptedProcessor.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/statemachine/runner/processors/VertexRunnerInterruptedProcessor.java
@@ -34,7 +34,6 @@ public class VertexRunnerInterruptedProcessor implements VertexRunnerPayloadProc
 
     @Override
     public void process(Throwable error) throws Exception {
-        logger.fine("jobManager.handleRunnerInterrupted " + vertexRunner.getVertex().getName());
         jobManager.handleRunnerInterrupted(error);
     }
 }

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/InterruptionTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/InterruptionTest.java
@@ -24,17 +24,18 @@ import com.hazelcast.jet.source.MapSource;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.TestHazelcastInstanceFactory;
 import com.hazelcast.test.annotation.QuickTest;
-import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicBoolean;
-import java.util.concurrent.locks.LockSupport;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Ignore;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.LockSupport;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
@@ -59,7 +60,6 @@ public class InterruptionTest extends JetTestSupport {
     }
 
     @Test
-    @Ignore //https://github.com/hazelcast/hazelcast-jet/issues/122
     public void testInterruptSlowApplication() throws Exception {
         int nodeCount = 2;
         HazelcastInstance instance = createCluster(factory, nodeCount);
@@ -72,12 +72,13 @@ public class InterruptionTest extends JetTestSupport {
         dag.addVertex(vertex);
         final Job job = JetEngine.getJob(instance, "testInterrupt", dag);
 
-        AtomicBoolean interrupted = new AtomicBoolean(false);
+        CountDownLatch interrupted = new CountDownLatch(1);
+
         new Thread(() -> {
             try {
                 SlowProcessor.latch.await();
                 job.interrupt().get();
-                interrupted.set(true);
+                interrupted.countDown();
             } catch (Exception e) {
                 e.printStackTrace();
             }
@@ -86,7 +87,29 @@ public class InterruptionTest extends JetTestSupport {
             job.execute().get();
             fail("The job was not interrupted");
         } catch (ExecutionException e) {
-            assertTrue(interrupted.get());
+            assertInstanceOf(InterruptedException.class, e.getCause());
+            assertOpenEventually(interrupted);
+        } finally {
+            job.destroy();
+        }
+    }
+
+    @Test
+    public void testInterruptAlreadyCompletedApplication() throws Exception {
+        int nodeCount = 2;
+        HazelcastInstance instance = createCluster(factory, nodeCount);
+        IMap<Integer, Integer> map = getMap(instance);
+        fillMapWithInts(map, COUNT);
+
+        DAG dag = new DAG();
+        Vertex vertex = createVertex("vertex", TestProcessors.Noop.class);
+        vertex.addSource(new MapSource(map));
+        dag.addVertex(vertex);
+        final Job job = JetEngine.getJob(instance, "testInterrupt", dag);
+
+        try {
+            job.execute().get();
+            job.interrupt().get();
         } finally {
             job.destroy();
         }
@@ -109,8 +132,8 @@ public class InterruptionTest extends JetTestSupport {
             execute(job);
             fail("The job should not execute successfully.");
         } catch (ExecutionException e) {
-            RuntimeException exception = (RuntimeException) e.getCause();
-            assertEquals(ExceptionProcessor.ERROR_MESSAGE, exception.getCause().getMessage());
+            Exception exception = (Exception) e.getCause();
+            assertEquals(ExceptionProcessor.ERROR_MESSAGE, exception.getMessage());
         }
     }
 
@@ -130,8 +153,8 @@ public class InterruptionTest extends JetTestSupport {
             execute(job);
             fail("The job should not execute successfully.");
         } catch (ExecutionException e) {
-            RuntimeException exception = (RuntimeException) e.getCause();
-            assertEquals(ExceptionProcessor.ERROR_MESSAGE, exception.getCause().getMessage());
+            Exception exception = (Exception) e.getCause();
+            assertEquals(ExceptionProcessor.ERROR_MESSAGE, exception.getMessage());
         }
     }
 

